### DISCUSSION TITLE
TLS handshake timeout

### DISCUSSION THREAD

**Discussion by tolstovskiy** [strongswan/strongswan#1606] [Q&A] [answered]
Hello there!

There is a tunnel between my Linux Box(Initiator) behind NAT and Cloud VM(Responder). I'm using Responder as sort of router with static public IP for Initiator.
At the moment Responder is successfully routing through tunnel all needed incoming traffic initiated from outside world to Initiator and also masquerading anything coming from Initiator.
But there is one problem with TLS handshake from Initiator to https://integrationapi.mango-office.ru/ (81.88.85.109). "Server Hello" packet from 81.88.85.109 is just not getting to Initiator.
I have a feeling it's something to do with length of the packet but I'm not savy enough to figure out what to do with it. Haven't encountered this problem with any other TLS handshake but I'm pretty sure the problem is on my 
side since i don't encounter this problem without tunnel and can replicate this problem on another Initiator from different network.

I've tried to mangle rules from https://docs.strongswan.org/docs/6.0/howtos/forwarding.html and just changing MTU of interface with ifconfig on both Initiator and Responder.

curl -v  https://integrationapi.mango-office.ru/ from Initiator:
```
*   Trying 81.88.85.109:443...
* Connected to integrationapi.mango-office.ru (81.88.85.109) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* SSL connection timeout
* Closing connection 0
```
Everywhere  below i've replaced Responder's public IP with 1.1.1.1.

swanctl.conf from Responder:
```
connections {
        ikev2 {
                version=2
                fragmentation=no
                dpd_delay=300s
                rekey_time=0
                local_addrs=%any
                remote_addrs=%any
                send_certreq=no
                proposals=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,3des-sha1-modp1024
#		pools=main-pool
                send_cert=always
                local {
                        id=1.1.1.1
                        certs=server-cert.pem
                }
                remote {
                        id=%any
                        eap_id=%any
                        auth=eap-mschapv2
                }
                children {
                        ikev2 {
                                start_action=none
                                ipcomp=no
                                mode=tunnel
                                dpd_action=clear
				remote_ts=10.0.0.0/8,172.16.0.0/12,192.168.0.0/16
                                local_ts=81.88.86.0/24,81.88.88.0/24,81.88.82.0/24,81.88.85.0/24,217.65.2.4,1.1.1.1/32
				esp_proposals=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1,3des-sha1
                        }
                }
        }
}
```

Here's output of "swanctl -l" from Responder:
```

ikev2: #7, ESTABLISHED, IKEv2, 8b65b8fdcbc7a42d_i 29539976ce841835_r*
  local  '1.1.1.1' @ 1.1.1.1[4500]
  remote 'mango_vpn' @ 81.88.83.41[4500]
  CHACHA20_POLY1305/PRF_HMAC_SHA2_512/CURVE_25519
  established 1371s ago
  ikev2: #9, reqid 1, INSTALLED, TUNNEL-in-UDP, ESP:AES_CBC-256/HMAC_SHA2_256_128
    installed 1371s ago, rekeying in 1914s, expires in 2589s
    in  ca15da53, 568968 bytes,  1383 packets,     1s ago
    out c62a0856, 706221 bytes,  1836 packets,     1s ago
    local  81.88.82.0/24 81.88.85.0/24 81.88.86.0/24 81.88.88.0/24 %1.1.1.1%/32 217.65.2.4/32
    remote 10.0.0.0/8 172.16.0.0/12 192.168.0.0/16

```
Here is my "iptables-save" output from responder with exclusion of fail2ban chains. I allowed pretty much everything at this point just in case I was droping something:
```
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
COMMIT

*filter
:INPUT ACCEPT [39351061:8402157126]
:FORWARD ACCEPT [525:208674]
:OUTPUT ACCEPT [2799560:751595906]
:
COMMIT
*nat
:PREROUTING ACCEPT [11166:940602]
:INPUT ACCEPT [2203:246844]
:OUTPUT ACCEPT [6:476]
:POSTROUTING ACCEPT [6:476]
-A PREROUTING -s 217.65.2.4/32 -d 1.1.1.1/32 -j DNAT --to-destination 192.168.27.123
-A PREROUTING -s 81.88.82.0/24 -d 1.1.1.1/32 -j DNAT --to-destination 192.168.27.123
-A PREROUTING -s 81.88.86.0/24 -d 1.1.1.1/32 -j DNAT --to-destination 192.168.27.123
-A PREROUTING -s 81.88.88.0/24 -d 1.1.1.1/32 -j DNAT --to-destination 192.168.27.123
-A PREROUTING -s 81.88.85.0/24 -d 1.1.1.1/32 -j DNAT --to-destination 192.168.27.123
-A POSTROUTING -s 192.168.27.123/32 -j MASQUERADE
-A POSTROUTING -s 192.168.200.210/32 -j MASQUERADE
COMMIT
```
Here's is also capture from Responder(Decoded):
[capture.txt](https://github.com/strongswan/strongswan/files/11000527/capture.txt)

Pretty sure anyone can copy my config and with some editing replicate the problem.
I'm hoping someone can at least give me a hint towards the solution of this problem. Thanks.

**Comment by tobiasbrunner** [comment, marked as answer]
> I have a feeling it's something to do with length of the packet but I'm not savy enough to figure out what to do with it.

Yes, definitely sounds like an MTU/MSS issue (also see the ICMP errors in the capture), please [refer to the docs](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_mtumss_issues).

**Reply by tolstovskiy** [reply]
I've tried this before but gave it another shot just in case. Maybe I'm doing it wrong? I'm setting everything to 1500 just as MTU of Responder's interface.

Here's my mangle table with additional rules.
```
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
-A FORWARD -p tcp -m policy --dir in --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1500 -j TCPMSS --set-mss 1500
-A FORWARD -p tcp -m policy --dir out --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1500 -j TCPMSS --set-mss 1500
```
kernel-netlink.conf from Responder excluding commented strings.
```

kernel-netlink {
    load = yes
     mss = 1500
     mtu = 1500
        ipv4 {
        }
        ipv6 {
        }
    }
}
```

**Reply by tobiasbrunner** [reply]
Yes, that's completely wrong. The whole point is that IPsec causes overhead that reduces the MTU/MSS for tunneled traffic from the default of 1500 to a much lower value (depends on the algorithms etc.).

**Reply by tolstovskiy** [reply]
Thanks, finally figured it out. Guess, I just needed someone to rephrase explanation from doc. Сonfirmation of my assumption also helped. I've have lowered MTU to 1350 for SYN,RST SYN packets coming from Initiator with mangle rule on Responder. For some reason it gets up to 1800 on Initiator's side but I'll figure out why on my own.
```
iptables -t mangle -A FORWARD -p tcp -m policy --dir in --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1350:1800 -j TCPMSS --set-mss 1350
iptables -t mangle -A FORWARD -p tcp -m policy --dir out --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1350:1800 -j TCPMSS --set-mss 1350
```
